src/Worker.Extensions.DurableTask/DurableTaskExtensionStartup.cs

```diff
@@ -26,13 +32,34 @@ public override void Configure(IFunctionsWorkerApplicationBuilder applicationBui
         applicationBuilder.Services.AddOptions<DurableTaskClientOptions>()
             .PostConfigure<IServiceProvider>((opt, sp) =>
             {
-                if (sp.GetService<DataConverter>() is DataConverter converter
-                    && !ReferenceEquals(opt.DataConverter, JsonDataConverter.Default))
+                if (GetConverter(sp) is DataConverter converter)
                 {
                     opt.DataConverter = converter;
                 }
             });
 
+        applicationBuilder.Services.AddOptions<DurableTaskWorkerOptions>()
+            .PostConfigure<IServiceProvider>((opt, sp) =>
+            {
+                if (GetConverter(sp) is DataConverter converter)
+                {
+                    opt.DataConverter = converter;
+                }
+            });
+
+        applicationBuilder.Services.TryAddSingleton(sp =>
+        {
+            DurableTaskWorkerOptions options = sp.GetRequiredService<IOptions<DurableTaskWorkerOptions>>().Value;
+            ILoggerFactory factory = sp.GetRequiredService<ILoggerFactory>();
+            return new DurableTaskShimFactory(options, factory); // For GrpcOrchestrationRunner
+        });
+
         applicationBuilder.UseMiddleware<DurableTaskFunctionsMiddleware>();
     }
+
+    private static DataConverter? GetConverter(IServiceProvider services)
+    {
+        WorkerOptions? worker = services.GetRequiredService<IOptions<WorkerOptions>>()?.Value;
+        return worker?.Serializer is ObjectSerializer serializer ? new ObjectConverterShim(serializer) : null;
```

Comment:
It looks like we only consider serializers that derive from `ObjectSerializer` in this code path. Can we/do we want to support users registering a `DataProvider` type directly?

Reply:
I intentionally left that out for now as providing a `DataConverter` directly is only a partial solution (it does not affect activities in durable functions?). When we solve that gap I am open to supporting direct `DataConverter`'s

Reply:
Added a comment explaining that.